mosec/manager.py

```diff
@@ -0,0 +1,309 @@
+        return method(self.pool)
+
+    def _start_process(
+        self,
+        worker_id: int,
+        stage_label: str,
+        work_path: str,
+        shutdown: Event,
+        shutdown_notify: Event,
+    ):
+        context = mp.get_context(self.start_method)
+        context = cast(Union[SpawnContext, ForkContext], context)
+        coordinator_process = context.Process(
+            target=Coordinator,
+            args=(
+                self.worker,
+                self.max_batch_size,
+                stage_label,
+                shutdown,
+                shutdown_notify,
+                work_path,
+                self.stage_id,
+                worker_id + 1,
+                self.ipc_wrapper,
+                self.timeout,
+            ),
+            daemon=True,
+        )
+
+        with env_var_context(self.env, worker_id):
+            coordinator_process.start()
+
+        self.pool[worker_id] = coordinator_process
+
+    def start(
+        self,
+        first: bool,
+        stage_label: str,
+        work_path: str,
+        shutdown: Event,
+        shutdown_notify: Event,
+    ) -> bool:
+        """Start the worker process.
+
+        Args:
+            first: whether the worker is tried to start at first time
+            stage_label: label of worker ingress and egress
+            work_path: path of working directory
+            shutdown: Event of server shutdown
+            shutdown_notify: Event of server will shutdown
+
+        Returns:
+            Whether the worker is started successfully
+        """
+        # for every sequential stage
+        self.pool = [p if self._process_healthy(p) else None for p in self.pool]
+        if self._healthy(all):
+            # this stage is healthy
+            return True
+        if not first and not self._healthy(any):
+            # this stage might contain bugs
+            return False
+
+        need_start_id = [
+            worker_id for worker_id in range(self.num) if self.pool[worker_id] is None
+        ]
+        for worker_id in need_start_id:
+            # for every worker in each stage
+            self._start_process(
+                worker_id, stage_label, work_path, shutdown, shutdown_notify
+            )
+        return True
+
+    def validate(self):
+        """Validate arguments of worker runtime."""
+        validate_env(self.env, self.num)
+        assert issubclass(
+            self.worker, Worker
+        ), "worker must be inherited from mosec.Worker"
+        validate_int_ge(self.num, "worker number")
+        validate_int_ge(self.max_batch_size, "maximum batch size")
+        validate_int_ge(self.max_wait_time, "maximum wait time", 0)
+        validate_int_ge(self.timeout, "forward timeout", 0)
+        assert (
+            self.start_method in NEW_PROCESS_METHOD
+        ), f"start method must be one of {NEW_PROCESS_METHOD}"
+
+
+class CoordinatorManager:
```

Comment:
Let's improve the naming to reduce confusion.

Maybe here we can call it `RuntimeManager` since it manages the runtime which further contains coordinator. Therefore, the original chain becomes:
```mermaid
graph LR
1[Server] --> 3[PyRuntimeManager]
1[Server] --> 2[RsRuntimeManager]
2[RsRuntimeManager] --> 4>Rust subprocess]
3[PyRuntimeManager] --> 5[Runtime]
5[Runtime] --> 8[Coordinator process]
8[Coordinator process] --> 9>num * Worker processes]
```

Here use `PyRuntimeManager` to distinguish `RsRuntimeManager`, and avoid naming too many chained abstractions.

cc @kemingy feel free to suggest any other naming options.

Reply:
Also, seems previously we use `Mosec process` to refer to the rust web server process. I feel that was not a good choice and it is not easy to understand. Perhaps we can just use "rust process" / "web layer process" to denote it.

Reply:
Thanks, the confused name is fixed.